Refactoring: Change `OverpassQuery` structure

holyturt opened the pull request:
This branch attempts to change `OverpassQuery` and its related classes to avoid user to make wrong query relationships. Especially, it:

- changes the way making instance of `OverpassQuery`,
```swift
// before
let node = SwiftOverpass.query(type: .node)
let way = node.related(.way)

// after
let node = NodeQuery()
let way = node.way()
```

- expresses relationships with mixins like below:
```swift
class NodeQuery: OverpassQuery, WayReferable, RelationReferable {
    // contains: way(_:), relation(_:)
}

class RelationQuery: OverpassQuery, NodeReferable, WayReferable, RelationReferable, 
 BackwardsReferable {
    // contains: node(_:), way(_:), relation(_:), backwards(_:)
}
```

- splits up methods which make xml query from `OverpassQuery` to new class `XMLQueryBuilder` to clarify these classes' concern and make it simpler.

This will have breaking changes and incompatibilities. Any ideas?

holyturt committed:
Change `OverpassQuery`'s structure and its dependents

holyturt committed:
Add basic query classes and mixins

holyturt committed:
Adapt the demo project to new `OverpassQuery`

holyturt committed:
Move query classes and related protocols to a dedicated group

holyturt committed:
Move `XMLQueryBuilder` to a dedicated group

holyturt committed:
Delete `UpQuery` for now

holyturt committed:
Delete `DownQuery` for now

holyturt committed:
Merge branch 'master' into refactor/query-builders

wtimme commented:
Hey @holyturt,

Great to see your changes! I was on vacation the last couple of weeks, and just saw the pull request. I'll be going through the files and add comments as I go along, just like I would during a code review. If you have any questions, suggestions or remarks, or if you have any ideas for the process, please comment.

wtimme commented on SwiftOverpass/Queries/OverpassQuery.swift:
```diff
@@ -0,0 +1,102 @@
+     - parameter w: Highest latitude
+     - parameter e: Highest longitude
+     */
+    public func setBoudingBox(s: Double, n: Double, w: Double, e: Double) {
```
To me as someone not (yet) using this part of the framework, it's a bit confusing to have two methods to set the bounding box. Is this for convenience reasons?

When setting the bounding box of a query, I'd write something along
```
let query = ...

query.boundingBox = BoundingBox(...)
```

Do you feel it's necessary to have these two methods? If we wanted to cover `OverpassQuery` with unit tests, we'd need to have one test for each of these, increasing the effort.

I suggest we remove these two methods and just use the
```
var boundingBox: BoudingBox? { get set }
```

holyturt replied:
Yes, you are right. It will make people confused. There are no reasons for those two methods. So I will edit the code along with your suggestion. Thanks!

wtimme commented on SwiftOverpass/Queries/QueryBuilder.swift:
```diff
@@ -0,0 +1,13 @@
+
+import Foundation
+
+public protocol QueryBuilder {
```
Nice job working with protocols! 👍

wtimme commented on SwiftOverpassDemo/SwiftOverpassDemo/ViewController.swift:
```diff
@@ -15,11 +15,12 @@ class ViewController: UIViewController {
         
-        let query = SwiftOverpass.query(type: .node)
-        query.hasTag("name", equals: "Schloss Neuschwanstein")
+        let node = NodeQuery()
```
Great! Not only is it shorter than the original approach, but it also prevents people from using the framework in a "wrong" way, reducing frustration. 🚀

wtimme commented on SwiftOverpass/SwiftOverpass.swift:
```diff
@@ -148,15 +118,16 @@ public final class OverpassApi {
     public func fetch(_ queries: [OverpassQuery], verbosity: OutputVerbosity? = nil, order: OutputOrder? = nil, completion: @escaping CompletionClosure) {
-        let requestQuery = makeXmlDocumentForRequestBody(queries: queries, verbosity: verbosity, order: order)
-        let parameters: [String : String] = ["data" : requestQuery.xmlCompact]
+        let builder = XMLQueryBuilder(queries: queries, verbosity: verbosity, order: order, timeout: timeout, elementLimit: elementLimit)
```
By moving this private functionality to a dedicated object, we will be able to improve the testability. Thanks!

wtimme reviewed (approved):
Looks good, thanks for your effort!

It would be great to have the `XMLQueryBuilder` covered by unit tests, but that is something we can easily add later on.

As for the API changes, I don't see any big effort involved when updating the pod.

holyturt committed:
Remove two unnecessary methods from `OverpassQuery`.

The `setBoundingBox()`s are confusing. Therefore, we use only `boundingBox` property to set bbox.

holyturt commented:
Hey, @wtimme!
Thank you for your review! That’s very helpful.
I will merge this PR with confidence!
